larrecodnn/ImagePatternAlgs/Tensorflow/Modules/EvaluateROIEff_module.cc

```diff
@@ -0,0 +1,151 @@
+  EvaluateROIEff(EvaluateROIEff const&) = delete;
+  EvaluateROIEff(EvaluateROIEff&&) = delete;
+  EvaluateROIEff& operator=(EvaluateROIEff const&) = delete;
+  EvaluateROIEff& operator=(EvaluateROIEff&&) = delete;
+
+  // Required functions.
+  void analyze(art::Event const& e) override;
+
+private:
+
+  void beginJob();
+  void endJob();
+  // Declare member data here.
+  TH1D *h_energy[3];
+  TH1D *h_energy_roi[3];
+  TEfficiency *eff_energy[3];
+
+  art::InputTag fWireProducerLabel;
+
+};
+
+
+nnet::EvaluateROIEff::EvaluateROIEff(fhicl::ParameterSet const& p)
+  : EDAnalyzer{p},
+  fWireProducerLabel(p.get<art::InputTag>("WireProducerLabel",""))
+  // More initializers here.
+{
+  // Call appropriate consumes<>() for any products to be retrieved by this module.
+}
+
+void nnet::EvaluateROIEff::analyze(art::Event const& e)
+{
+
+  auto const* geo = lar::providerFrom<geo::Geometry>();
+  auto const* detprop = lar::providerFrom<detinfo::DetectorPropertiesService>();
+  auto const* tclk = lar::providerFrom<detinfo::DetectorClocksService>();
+  auto const& chStatus = art::ServiceHandle< lariov::ChannelStatusService >()->GetProvider();
+
+  art::Handle < std::vector < recob::Wire > > wireListHandle;
+  std::vector < art::Ptr < recob::Wire > > wires;
+  if (e.getByLabel(fWireProducerLabel, wireListHandle)) {
+    art::fill_ptr_vector(wires, wireListHandle);
+  }
+
+  auto simChannelHandle = e.getValidHandle<std::vector<sim::SimChannel>>("tpcrawdecoder:simpleSC");
+
+  // ... Loop over simChannels
+  for ( auto const& channel : (*simChannelHandle) ){
+    
+    // .. get simChannel channel number
+    const raw::ChannelID_t ch1 = channel.Channel();
+    if (chStatus.IsBad(ch1)) continue;
+    if (ch1%1000 == 0) std::cout<<ch1<<std::endl;
+    int view = geo->View(ch1);
+    auto const& timeSlices = channel.TDCIDEMap();
+    for ( auto const& timeSlice : timeSlices ){
+
+      auto const& energyDeposits = timeSlice.second;
+      auto const tpctime = timeSlice.first;
+      int tdctick = static_cast<int>(tclk->TPCTDC2Tick(double(tpctime)));
+      //if(tdctick!=tpctime)std::cout << "tpctime: " << tpctime << ", tdctick: " << tdctick << std::endl;
+      if(tdctick<0 || tdctick>int(detprop->ReadOutWindowSize())-1) continue;
+      double totalE = 0;
+      for ( auto const& energyDeposit : energyDeposits ){
+        totalE += energyDeposit.energy;
+      }
+      //std::cout<<totalE<<std::endl;
+      h_energy[view]->Fill(totalE);
+      for (auto & wire : wires){
+        if (wire->Channel() != ch1) continue;
+        const recob::Wire::RegionsOfInterest_t& signalROI = wire->SignalROI();
+        for(const auto& range : signalROI.get_ranges()){
+          const auto& waveform = range.data();
+          raw::TDCtick_t roiFirstBinTick = range.begin_index();
+          if (tdctick>=roiFirstBinTick && tdctick < int(roiFirstBinTick + waveform.size())){
+            h_energy_roi[view]->Fill(totalE);
+            break;
```

Comment:
This breaks out of only the innermost loop (l. 115).  Is that the intention?

Reply:
This is correct.